Add euler conversion from Quaternion

Melix19 opened the pull request:
I think that this method could be helpful in some situation using magnum, so I decided to make a PR.
I took the algorithm from [QT](https://github.com/qt/qtbase/blob/84056cdca11c63ddb9fac3d95fcfa952adaa6ea1/src/gui/math3d/qquaternion.cpp#L507) (that it's based on another algorithm as specified in their comments) and adapted to magnum using a Vector3 as return.

It's obvious that this method should not be abused because of the problems that eulers can have (and I think that this type of conversion isn't even always 100% perfect), so maybe we could specify that in the doc?

mosra commented on src/Magnum/Math/Quaternion.h:
```diff
@@ -399,6 +399,13 @@ template<class T> class Quaternion {
+         *
+         * @see @ref rotation(), @ref DualQuaternion::rotation()
+         */
+        Vector3<T> toRotation() const;
```
`toEuler()` instead? So it's clear that you're going to the not-recommended API areas :) Can it return `Vector3<Rad<T>>`? Would be good to mention in which order the returned values are, in which order they should be combined and also a snippet that shows how to convert the Euler angles back:

```cpp
Rad x, y, z;
Quaternion q = /* I hope this is correct, please check! */
    Quaternion::rotate(z, Vector3::zAxis())*
    Quaternion::rotate(y, Vector3::yAxis())*
    Quaternion::rotate(x, Vector3::xAxis());
```

To avoid syntax error in the documentation, the docs have compiled snippets, so instead of placing it directly here what would be great is to extend `doc/snippets/MagnumMath.cpp` with a block delimited by `/* [Quaternion-toRotation] */` and reference it via `@snippet MagnumMath.cpp Quaternion-toRotation` (check other places, e.g. in Matrix4.h, to see how it's usually done). Of course if you don't feel like doing that, I can do that post-merge (but that'll delay the merge, of course ... I'm a bit tight with time lately).

mosra commented on src/Magnum/Math/Quaternion.h:
```diff
@@ -733,6 +740,57 @@ template<class T> Matrix3x3<T> Quaternion<T>::toMatrix() const {
 }
 
+/* Algorithm from:
+   https://github.com/qt/qtbase/blob/84056cdca11c63ddb9fac3d95fcfa952adaa6ea1/src/gui/math3d/qquaternion.cpp#L507 */
```
Could this reference the original code Qt is based off, instead? Are there any differences? Because Qt is LGPL and GPL means problems :)

mosra commented on src/Magnum/Math/Quaternion.h:
```diff
@@ -733,6 +740,57 @@ template<class T> Matrix3x3<T> Quaternion<T>::toMatrix() const {
+/* Algorithm from:
+   https://github.com/qt/qtbase/blob/84056cdca11c63ddb9fac3d95fcfa952adaa6ea1/src/gui/math3d/qquaternion.cpp#L507 */
+template<class T> Vector3<T> Quaternion<T>::toRotation() const {
+    Vector3<T> euler;
```
```suggestion
    Vector3<T> euler{NoInit};
```

mosra commented on src/Magnum/Math/Quaternion.h:
```diff
@@ -733,6 +740,57 @@ template<class T> Matrix3x3<T> Quaternion<T>::toMatrix() const {
+        }
+    } else {
+        euler.z() = 0.0f;
+        euler.y() = std::atan2(-2.0f*(xy - zw), 1.0f - 2.0f*(yy + zz));
```
All the literals need to be `T(0.0)` and `T(2.0)` respectively to avoid compiler warnings when using the double-precision `Quaterniond`.

mosra commented on src/Magnum/Math/Quaternion.h:
```diff
@@ -733,6 +740,57 @@ template<class T> Matrix3x3<T> Quaternion<T>::toMatrix() const {
+
+    const T lengthSquared = xx + yy + zz + wp*wp;
+    if(!TypeTraits<T>::equalsZero(lengthSquared - 1.0f, 100) &&
+        !TypeTraits<T>::equalsZero(lengthSquared, 100)) {
```
Not sure about the `100` here -- doesn't it make the comparison *too* fuzzy? If the quat is normalized (which is a prerequisite for extracting the angles?), the original magnitude is `T(1.0)`, so that's what should be there I think.

mosra commented on src/Magnum/Math/Quaternion.h:
```diff
@@ -733,6 +740,57 @@ template<class T> Matrix3x3<T> Quaternion<T>::toMatrix() const {
 
+/* Algorithm from:
+   https://github.com/qt/qtbase/blob/84056cdca11c63ddb9fac3d95fcfa952adaa6ea1/src/gui/math3d/qquaternion.cpp#L507 */
+template<class T> Vector3<T> Quaternion<T>::toRotation() const {
```
Shouldn't this assert for the quaternion being normalized, similarly to how `angle()` / `axis()` have it? AFAIK Qt's quaternions are always normalized, so there they don't check this.

The docs should then mention 

> Expects the quaternion to be normalized.

Melix19 commented:
Finally had time for this!

So, I changed the algorithm to the one found in [wikipedia](https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles#Quaternion_to_Euler_Angles_Conversion), the reason is that the Qt's one was giving a "dancing" result when converting from/to eulers in every frame (I know that this is never a good idea anyway, but it was just for testing):
![Dancing rect](https://user-images.githubusercontent.com/23530586/69188914-b7925880-0b1d-11ea-92d9-b5d04e30d955.gif)

With this new algorithm is basically perfect.
Then, for the tests I just added a random quaternion and compared the euler result to a result given from [this online converter](https://www.andre-gaschler.com/rotationconverter/) given the same quaternion.

If there's nothing else to do, I think that it can be merged :D

alanjfs commented:
Woho! Good timing! I also need this. :D

mosra commented on src/Magnum/Math/Quaternion.h:
```diff
@@ -733,6 +746,39 @@ template<class T> Matrix3x3<T> Quaternion<T>::toMatrix() const {
+    /* Pitch (y-axis rotation) */
+    T sinp = T(2.0) * (qw * qy - qz * qx);
+    if(std::abs(sinp) >= T(1.0))
+        euler.y() = Rad<T>(std::copysign(Constants<T>::pi() / T(2.0), sinp)); /* Use 90 degrees if out of range */
```
```suggestion
        euler.y() = Rad<T>(std::copysign(Constants<T>::piHalf(), sinp)); /* Use 90 degrees if out of range */
```

Also, codecov is reporting this line is not covered. Can you craft a value for the test that hits this condition? :) Then it'll be perfect.

Melix19 commented:
Sorry for the delay.
I added another test with other values trying to cover the remaining line, but for some reason it gives me the wrong result. For reference:

```
    Quaternion a2 = Quaternion({0.70f, 0.0f, 0.70f}, 0.0f).normalized();
    Math::Vector3<Rad> b2{0.0_radf, -1.57079_radf, -3.14159_radf};

    CORRADE_COMPARE(a2.toEuler(), b2);
```
```
    Values a2.toEuler() and b2 are not the same, actual is
    Vector(Rad(0), Rad(-1.57045), Rad(0))
    but expected
    Vector(Rad(0), Rad(-1.57079), Rad(-3.14159))
```

So, I changed the algorithm again, using directly the one that [this](https://www.andre-gaschler.com/rotationconverter/) online converter [uses](https://github.com/mrdoob/three.js/blob/6892dd0aba1411d35c5e2b44dc6ff280b24d6aa2/src/math/Euler.js#L197) that in my tests it seems the perfect one. The licence is MIT, so it should be good.

Hopefully the coverage is 100% now.

Melix19 commented on src/Magnum/Math/Quaternion.h:
```diff
@@ -733,6 +746,37 @@ template<class T> Matrix3x3<T> Quaternion<T>::toMatrix() const {
+    T m23 = rotMatrix[1][2];
+    T m33 = rotMatrix[2][2];
+
+    euler.y() = Rad<T>(std::asin(-Math::min(Math::max(m13, T(-1.0)), T(1.0))));
```
Oh, and here I tried to use the `Math::clamp()` instead of the `Math::min(Math::max())`, but for some reason it says that it does not exist (?)

mosra replied:
It's defined in `Functions.h` that are not included here. This is fine :)

mosra commented:
Merged as 097289850aeebb9283e20de7b060d9585457c2e5, thank you! Interesting that there's so many algos out there but only some of them work as expected :D
